services/aster-nix/src/fs/file_table.rs

```diff
@@ -127,6 +135,29 @@ impl FileTable {
         closed_files
     }
 
+    pub fn close_files_on_exec(&mut self) -> Vec<Arc<dyn FileLike>> {
+        let mut closed_files = Vec::new();
+        let closed_fds: Vec<FileDescripter> = self
+            .table
+            .idxes_and_items()
+            .filter_map(|(idx, entry)| {
+                if entry.is_cloexec() {
+                    Some(idx as FileDescripter)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        for fd in closed_fds {
+            let entry = self.table.remove(fd as usize).unwrap();
+            let events = FdEvents::Close(fd);
+            self.notify_fd_events(&events);
+            entry.notify_fd_events(&events);
```

Comment:
I think `FileTable` should avoid calling `notify_fd_events`, which triggers user-defined callbacks, while holding a `&mut self`. Having `&mut self` means having a `MutuxGuard<FileTable>` . So what if the user-defined callbacks also need to access the file table? This would lead to deadlock. This problem is not unique to `close_files_on_exec`; other methods suffer a similar problem.

In general, invoking user-defined callbacks while holding a lock raises red flags. A recipe for deadlock.

Reply:
Maybe modifying this struct to an internal mutable(`Mutex<SlotVec<FileTableEntry>>`) struct can solve this problem.

Reply:
Submit a new issue #663 to track this problem.